code/modules/projectiles/guns/projectile/CMturret.dm

```diff
@@ -0,0 +1,223 @@
+	icon_state = "mgun+barrier"
+	var/fire_sound = 'sound/weapons/Gunshot_smg.ogg'
+	var/empty_sound = 'sound/weapons/empty.ogg'
+	var/ammo_type = "/obj/item/projectile/bullet/a762"
+	var/ammo = 500
+	var/ammomax = 500
+	var/list/row1 = list()
+	var/list/row2 = list()
+	var/list/row3 = list()
+	var/mob/living/carbon/human/user
+	var/FIRETIME = 1 //tenths of seconds
+	density = 1
+	anchored = 0
+	flags = ON_BORDER
+	var/health = 500
+
+/obj/structure/machinegun/process()
+	var/mob/living/carbon/human/U
+	for(var/mob/living/carbon/human/H in range(0, src))
+		if(H)
+			U = H
+	if(U)
+		if(user && user != U)
+			user.mounted = null
+			user = null
+		user = U
+		user.mounted = src
+	else
+		if(user)
+			user.mounted = null
+			user = null
+
+/obj/structure/machinegun/New()
+	..()
+	processing_objects.Add(src)
+
+/obj/structure/machinegun/proc/shoot(var/turf/T)
+	if(ammo <= 0)
+		if(user)
+			playsound(src, empty_sound, 70, 1)
+			to_chat(user, "This machine gun is out of ammo!")
+
+		return
+	if(T && user && !user.incapacitated())
+		var/row = 0
+		if(row1.Find(T))
+			row = 1
+		else if(row2.Find(T))
+			row = 2
+		else if(row3.Find(T))
+			row = 3
+		if(row)
+			var/turf/shootfrom
+			switch(row)
+				if(1)
+					shootfrom = get_step(src, turn(dir, 90))
+				if(2)
+					shootfrom = get_step(src, dir)
+				if(3)
+					shootfrom = get_step(src, turn(dir, 270))
+			if(shootfrom)
+
+				var/turf/curloc = get_turf(shootfrom)
+				var/turf/targloc
+				switch(row)
+					if(1)
+						targloc = row1[7]
+					if(2)
+						targloc = row2[7]
+					if(3)
+						targloc = row3[7]
+				if (!istype(targloc) || !istype(curloc))
+					return
+				playsound(src, fire_sound, 50, 1)
+				var/obj/item/projectile/bullet/a762/bullet = new ammo_type(shootfrom)
+				bullet.firer = user
+				bullet.def_zone = user.zone_sel.selecting
+				bullet.original = T
+				bullet.loc = get_turf(shootfrom)
+				bullet.starting = get_turf(shootfrom)
+				bullet.shot_from = src
+				bullet.silenced = 0
+				bullet.current = curloc
+				bullet.yo = targloc.y - curloc.y
+				bullet.xo = targloc.x - curloc.x
+				ammo--
+				spawn()
+					if(bullet)
+						bullet.process()
+				user.changeNext_move(CLICK_CD_MOUNTED)
+
+
+/obj/structure/machinegun/proc/update_rows()
+	row1 = list()
+	row2 = list()
+	row3 = list()
+	//row1 + 90
+	//row2 + 0
+	//row3 - 90
+
+	//row1
+	var/turf/pos = get_step(src, dir)
+	pos = get_step(pos, turn(dir, 90))
+	row1.Add(pos)
+	var/i
+	for(i=2, i<8, i++)
+		row1.Add(get_turf(get_step(row1[i - 1], dir)))
+	//row2
+	pos = get_step(src, dir)
+	row2.Add(pos)
+	for(i=2, i<8, i++)
+		row2.Add(get_turf(get_step(row2[i - 1], dir)))
+	//row3
+	pos = get_step(src, dir)
+	pos = get_step(pos, turn(dir, 270))
+	row3.Add(pos)
+	for(i=2, i<8, i++)
+		row3.Add(get_turf(get_step(row3[i - 1], dir)))
+
+
+/obj/structure/machinegun/attackby(obj/item/W as obj, mob/user as mob)
+	var/obj/item/machinegunammo/Ammo = W
+	if(istype(W, /obj/item/weapon/wrench))
```

Comment:
Everything from here to line 155 should be replaced with `default_unfasten_wrench(user, W, 40)`. You'll also likely want to overload the proc for machinegun so that it also calls update_rows() when needed.